### ISSUE TITLE
decodeURIComponent can throw at runtime

### ISSUE THREAD

**Issue by delvedor** [yoshuawuyts/wayfarer#57]
Hello!
First of all, thank for this module, I love it.

I was testing a route and I mistyped the url, for instance instead of write `2` I wrote `"`, the result was that [decodeURIComponent](https://github.com/yoshuawuyts/wayfarer/blob/master/trie.js#L81) threw an `URIError` exception at runtime!
I can easily fix this by using a try catch and return `undefined`, but it will deoptimize the function.
We can also use a regex, something like:
```js
if (/"/g.test(thisRoute)) return undefined
```
But actually I don't know if `"` is the only bad character!
Any suggestion? :)

Thanks!

**Comment by yoshuawuyts** [comment]
Do you have an example input? I'm not sure how to reproduce this

**Comment by delvedor** [comment]
Sure!
Server:
```js
'use strict'

const http = require('http')
const wayfarer = require('wayfarer')

const router = wayfarer()
router.on('/test', (params, req, res) => {
  res.end()
})

router.on('/test/:id', (params, req, res) => {
  res.end()
})

const server = http.createServer((req, res) => {
  router(req.url, req, res)
})

server.listen(3030, err => {
  if (err) throw err
  console.log('server started')
})
```

Dangerous url:
```
http://localhost:3030/test/cia%"Fo
```

I've made some test and I discovered that if I use the `"` between some letters, `cia"o` for example everything is ok, but if I put the `"` between a url encoded value, such as `/` which is `%2F`, *decodeURIComponent* will throw.
Surely this is not your fault, it is an edge case related to the *decodeURIComponent* implementation.

**Comment by yoshuawuyts** [comment]
@delvedor uh oh, I see what you mean now. This is really not good (ddos potential is, well, very real now) What do you suggest is the best course of action here?

Also cc/ @mcollina, feel we're talking perf and security for Fastify now too, I think you might have some input on the matter maybs?

**Comment by delvedor** [comment]
Sanitize the url is not the best solution, since we will remove some characters, as you seen in my example, the `"` will not always throw.
The easiest solution is to use a `try catch` in an external function, to avoid performances regressions.

cc also @davidmarkclements

**Comment by davidmarkclements** [comment]
Oh wow, nice catch. `decodeURIComponent` is surely used everywhere (I'm sure I've seen it all over the place in Hapi source for instance).

Questions

Does `decodeURI` do the same thing

What happens if you do `decodeURIComponent(JSON.stringify(uri))`



Re the `try/catch` - it's optimized in recent v8 versions (I've tested in Node 6, it seems to have landed) we might be ok to not even wrap in a function

**Comment by davidmarkclements** [comment]
wait - try `decoreURIComponent(escape(uri))`

**Comment by davidmarkclements** [comment]
```
> decodeURIComponent(escape('http://localhost:3030/test/cia%"Fo'))
'http://localhost:3030/test/cia%"Fo'
```

:bowtie:

**Comment by davidmarkclements** [comment]
be interesting to see perf differences of different approaches, reckon `escape` will be fastest though

**Comment by delvedor** [comment]
@davidmarkclements `decodeURI` throws in the same way.

Benchmarks:
```
decodeURIComponent x 3,835,931 ops/sec ±0.91% (92 runs sampled)
decodeURIComponent(escape(uri)) x 2,080,360 ops/sec ±0.70% (93 runs sampled)

decodeURI x 4,213,128 ops/sec ±0.76% (92 runs sampled)
decodeURI(escape(uri)) x 2,106,648 ops/sec ±0.67% (92 runs sampled)
```

**Comment by delvedor** [comment]
@davidmarkclements use `decodeURIComponent ` and `escape` together is not a good solution, since it introduces a wrong behaviour:
```js
decodeURIComponent(escape('http://localhost:3030/test/cia%2Fo'))
// => 'http://localhost:3030/test/cia%2Fo'

decodeURIComponent('http://localhost:3030/test/cia%2Fo')
// => 'http://localhost:3030/test/cia/o'
```

Probably the best solution is to use a `try/catch`
```js
try {
  decodeURIComponent('http://localhost:3030/test/cia%2Fo')
} catch (e) {
  return undefined
}
```
Here the benchmarks (in node `v6.10.0`) within and without the `try/catch`:
```
decodeURIComponent - no try/catch x 3,999,819 ops/sec ±0.79% (91 runs sampled)
decodeURIComponent - try/catch x 3,842,394 ops/sec ±0.70% (93 runs sampled)
```

**Comment by yoshuawuyts** [comment]
Looks like introducing try/catch doesn't introduce a huge performance
regression and solves the issue. Could you open a PR for this?

**Comment by delvedor** [comment]
Closed by #60 🎉
